### ISSUE TITLE
Which distance formula and box co-ordinates or delta is used to compute DIOU loss

### ISSUE THREAD

**Issue by jfern-1084** [Zzh-tju/DIoU-pytorch-detectron#14]
Thank you for providing the reference code.

I am just a bit confused and would like some clarification on what distance formula was used here?
https://github.com/Zzh-tju/DIoU-pytorch-detectron/blob/6e18f2c9f80c995e8730605de5aaabfa346e88d0/lib/utils/net.py#L70

As per your paper Euclidean distance was used to compute distance between center points of boxes. But I didn't notice a square root in the calculation.

Another concern I have is about whether delta values are used to compute the loss or if box co-ordinates are used. In Detectron2 and MMDetection they have used box co-ordinates and not delta values. As per your code it looks like delta values are used. Have you tried feeding box co-ordinates to the loss functions?

**Comment by Zzh-tju** [comment]
The reason why we did not use Euclidean distance is due to the gradient calculation. Note that if DIoU_term= sqrt(d²/c²), then the gradient of x will be 1/(2*sqrt(d²/c²))*∂(d²/c²)/∂x. Since the optimization is to try to minimize d/c, the existance of 1/(2*sqrt(d²/c²)) will enlarge the gradient.

As for using delta value, not box coordinates, we just follow GIoU paper, and everything is ok so far. We did not observe any explosion during training and can get a normal results. You can transform to box coordinate to see what will happen.

**Comment by jfern-1084** [comment]
Thank you for providing the reasoning behind not using the square root and for using delta values instead of box co-ordinates. Really appreciate the detailed explanation.

I was hoping, if you could also help me understand how to pick **regression loss weight** to normalize the regression and classification loss. In your paper you picked 12 while in the GIOU paper they picked 10. I have two test datasets and one works with well 10 while the other works with 12. I'm not sure how to pick the right value.

For instance I have some iterations where with weight 10, I get classification_loss as 0.015 and regression_loss as 0.045. For all iterations, the regression loss is slightly higher than classification loss as shown in this example. Would this be considered a normalized value or would you recommend increasing the weight?

**Comment by Zzh-tju** [comment]
We have observed that increasing the bbox reg weight properly can improve the performance. And similar arguments also appeared in Libra R-CNN and GHM.

For Faster/Mask R-CNN, I recommend you to use regression weight 12  rather than 10.  But ＞12 may cause unstable training. 

For SSD, I recommend to  regression weight 16. You can refer to (https://github.com/Zzh-tju/DIoU-SSD-pytorch)

**Comment by jfern-1084** [comment]
Yes, as you have pointed out, I have witnessed regression weight > 12 is definitely quite unstable. I am working with Faster R-CNN so I will stick to 12 as you've mentioned.

Thank you for your recommendations.
